### PR TITLE
PRS Scoring Pipeline for BGE

### PR DESCRIPTION
**Pull request by michaelgatzen**
This PR adds two WDLs: ProGRESSMultivariateRiskModel (along with changes made in that branch) and ScoreBGE.

Additionally, it adds a docker "python-data-slim-pysam" to the palantir-workflow docker library, which the ScoreBGE docker builds upon.

**Main file to focus on:** `ImputationPipeline/ScoreBGE/ScoreBGE.py`

### CODE REVIEW

File: ImputationPipeline/ScoreBGE/ScoreBGE.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+                for weight in self.prs_weights.iterrows():
+                    locus = weight[1]['locus']
+                    ref = weight[1]['ref']
+                    alt = weight[1]['alt']
+                    weight_samples = [sample_name for sample_name in self.sample_names if (locus, ref, alt) in self.gvcf_sites_scored[sample_name]]
+                    out_exome_gvcf_sites_scored.write(f'{locus}:{ref}:{alt}\t{",".join(weight_samples)}\n')
+
+        # VCF
+        if self.vcf_sites_scored is None:
+            if not allow_single_source_scoring:
```
Comment (kachulis) [inline ImputationPipeline/ScoreBGE/ScoreBGE.py:241]: could refactor this a little, since it's pretty identical to the gvcf code, but not a big deal either way.

File: ImputationPipeline/ScoreBGE/ScoreBGE.wdl [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    command <<<
+        set -xeuo pipefail
+        
+        cat <<'EOF' > script.py
+import sys
+sys.path.insert(0, '/ScoreBGE')
+import ScoreBGE
+import argparse
+
+parser = argparse.ArgumentParser(description='Score BGE')
```
Comment (kachulis) [inline ImputationPipeline/ScoreBGE/ScoreBGE.wdl:82]: Personally, I think you could move all of this into `ScoreBGE.py`, and just have this task be `python3 ScoreBGE.py ...`.  You'd need to add the other inputs to your arg parser, but I think it'd be worth it.  Then you can avoid the `sys.path.insert` as well.

File: ImputationPipeline/ScoreBGE/tests/ScoreBGE_test.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
+import sys
```
Comment (kachulis) [inline ImputationPipeline/ScoreBGE/tests/ScoreBGE_test.py:1]: any reason not to use one of the python testing frameworks for this?

File: ImputationPipeline/ScoreBGE/tests/ScoreBGE_test.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+sys.path.insert(0, '/Users/mgatzen/code/palantir-workflows')
```
Comment (kachulis) [inline ImputationPipeline/ScoreBGE/tests/ScoreBGE_test.py:2]: there must be a way to make this work without including this

File: ImputationPipeline/ScoreBGE/tests/ScoreBGE_test.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        #                               ('1:100', 'A', 'C'), no call low gq
+        ('1:200', 'A', 'C'), # score 0
+        ('1:300', 'A', 'C'), # score 4
+        ('1:310', 'A', 'AC'), #score 0
+        ('1:320', 'AC', 'A'), #score 16
+        #                               ('1:400', 'A', 'C'), negative GQ
+        ('1:500', 'A', 'C'), # score 64
+        ('1:700', 'G', 'C'), # score 0
+        #                               ('1:800', 'A', 'C'), gap
+        ('1:900', 'A', 'C'), # score 1024
```
Comment (kachulis) [inline ImputationPipeline/ScoreBGE/tests/ScoreBGE_test.py:50]: add a test for the gvcf score calculation that scores some non-ref sites (het, hom-var) with the effect allele being both ref and alt.

also would be good to have same tests for the wgs scoring, and maybe a simple test for the combination.

File: ImputationPipeline/ScoreBGE/tests/resources/test_ref_blocks.gvcf [hunk 6]
```diff
@@ -0,0 +1,60 @@
+##fileformat=VCFv4.2
+##ALT=<ID=NON_REF,Description="Represents any possible alternative allele at this location">
+##FORMAT=<ID=AD,Number=R,Type=Integer,Description="Allelic depths (counting only informative reads out of the total reads) for the ref and alt alleles in the order listed">
```
Comment (kachulis) [inline ImputationPipeline/ScoreBGE/tests/resources/test_ref_blocks.gvcf:3]: clean up vcf, shouldn't need all these header lines just for testing

File: ImputationPipeline/ScoreBGE/tests/resources/test_ref_blocks.gvcf [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+##FORMAT=<ID=SB,Number=4,Type=Integer,Description="Per-sample component statistics which comprise the Fisher's Exact Test to detect strand bias">
+##FORMAT=<ID=SPL,Number=.,Type=Integer,Description="Normalized, Phred-scaled likelihoods for SNPs based on the reference confidence model">
+##FORMAT=<ID=SQ,Number=1,Type=Float,Description="Somatic quality">
+##DRAGENCommandLine=<ID=HashTableBuild,Version="SW: 01.003.044.3.7.0-836-g2c77f4e2, HashTableVersion: 8",CommandLineOptions="dragen --build-hash-table true --ht-build-rna-hashtable true --ht-reference /illumina/scratch/DRAGEN/data/vault/reference_genomes/Hsapiens/hg38/seq/hg38.fa --output-directory /staging/tmp//illumina/scratch/DRAGEN/users/rruuspakka/rebuild_hash_tables/135944/EId5/hg38_alt_aware_graph_based/DRAGEN/8 --ht-num-threads 40 --ht-alt-liftover /opt/edico/liftover/bwa-kit_hs38DH_liftover.sam --enable-cnv true --ht-pop-alt-contigs /opt/edico/liftover/pop_altContig.fa.gz --ht-pop-alt-liftover /opt/edico/liftover/pop_liftover.sam.gz --ht-pop-snps /opt/edico/liftover/pop_snps.vcf.gz">
```
Comment (kachulis) [inline ImputationPipeline/ScoreBGE/tests/resources/test_ref_blocks.gvcf:20]: definitely don't need these

File: ImputationPipeline/ScoringTasks.wdl [hunk 8]
```diff
@@ -22,9 +24,9 @@ task ScoreVcf {
 
   command <<<
     /plink2 --score ~{weights} header ignore-dup-ids list-variants no-mean-imputation \
-    cols=maybefid,maybesid,phenos,dosagesum,scoreavgs,scoresums --set-all-var-ids ~{var_ids_string} --allow-extra-chr ~{extra_args} -vcf ~{vcf} dosage=DS \
-    --new-id-max-allele-len 1000 missing ~{"--extract " + sites} --out ~{basename} --memory ~{plink_mem} ~{"--output-chr " + chromosome_encoding}
-  >>>
+    cols=maybefid,maybesid,phenos,dosagesum,scoreavgs,scoresums --set-all-var-ids ~{var_ids_string} --allow-extra-chr ~{extra_args} -vcf ~{vcf} ~{if use_dosage_annotation then "dosage=DS" else ""} \
+    --new-id-max-allele-len 1000 missing ~{"--extract " + sites} ~{"--exclude " + exclude_sites} --out ~{basename} --memory ~{plink_mem} ~{"--output-chr " + chromosome_encoding}
+  }
```
Comment (kachulis) [inline ImputationPipeline/ScoringTasks.wdl:29]: are changes to this file needed?  and right now this looks like it would break the wdl.

File: ImputationPipeline/ScoreBGE/tests/ScoreBGE_test.py (hunk 5 above)
Comment (kachulis) [reply ImputationPipeline/ScoreBGE/tests/ScoreBGE_test.py:50]: (I think there may be a bug related to the effect allele index that more tests would catch)

**Review by kachulis — CHANGES_REQUESTED** [review]
Thanks @michaelgatzen!  A few requests, but looks generally in very good shape.  Main things are whether there is an off by 1 error related to the effect allele index when scoring a gvcf, and a request for a bit more tests of the scoring since you are rewriting it completely in python.

File: ImputationPipeline/ProGRESSMultivariateRiskModel.wdl [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+            ref_fasta = ref_fasta,
+            ref_fasta_index = ref_fasta_index,
+            ref_dict = ref_dict
+    }
+
+
+    call PCATasks.ArrayVcfToPlinkDataset {
+        input:
+			vcf = imputed_wgs_vcf,
```
[the remark this answers is not in the document]
Comment (michaelgatzen) [reply ImputationPipeline/ProGRESSMultivariateRiskModel.wdl:53]: fixed

File: ImputationPipeline/ScoreBGE/ScoreBGE.py [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+        self.vcf_num_sites_not_found = None
+
+    def _read_ref_dict(self, ref_dict_path):
+        with open(ref_dict_path) as ref_dict:
+            return [line.split('\t')[1].replace('SN:', '') for line in ref_dict if line.startswith('@SQ')]
+
+    def _read_prs_weights(self, prs_weights_path):
+        prs_weights = pd.read_table(prs_weights_path, dtype={'effect_allele': str, 'weight': np.float64, 'chr': str, 'bp': np.int32, 'ref': str, 'alt': str})
+
+        prs_weights[['contig', 'position', 'ref', 'alt']] = prs_weights['CHR:BP:REF:ALT'].str.split(':', expand=True)
```
[the remark this answers is not in the document]
Comment (michaelgatzen) [reply ImputationPipeline/ScoreBGE/ScoreBGE.py:34]: Agreed. I tried to make the pipeline compatible with the current weights files but that might not be the best plan. I now defined a hopefully reasonable format and uploaded the PCARE weights to the bge-resources bucket

File: ImputationPipeline/ScoreBGE/ScoreBGE.py [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+        prs_weights = prs_weights.drop(['CHR:BP:REF:ALT'], axis=1)
+        prs_weights['locus'] = prs_weights['contig'] + ':' + prs_weights['position']
+
+        prs_weights = prs_weights.astype({'effect_allele': str, 'weight': np.float64, 'contig': str, 'position': np.int32, 'ref': str, 'alt': str})
```
[the remark this answers is not in the document]
Comment (michaelgatzen) [reply ImputationPipeline/ScoreBGE/ScoreBGE.py:38]: fixed

File: ImputationPipeline/ProGRESSMultivariateRiskModel.wdl [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+        File pc_meansd
+        File pc_sites
+
+        Float prs_beta
+        Float fam_hist_beta
+        Float pc1_beta
+        Float pc2_beta
+
+        Boolean use_ref_alt_for_ids = true
+        String chromosome_encoding = "chrMT"
```
[the remark this answers is not in the document]
Comment (michaelgatzen) [reply ImputationPipeline/ProGRESSMultivariateRiskModel.wdl:28]: Yes I could, although I was thinking given that this is a pipeline with a singular purpose, we could require a defined format rather than running that task (or call caching it) every time. Let me know what you think

File: ImputationPipeline/ScoreBGE/ScoreBGE.py [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+    def _compare_record_and_weight(self, record, weight):
+        if record.contig != weight.contig:
+            return self.ref_dict.index(record.contig) - self.ref_dict.index(weight.contig)
+        if record.pos <= weight.position <= record.stop:
+            return 0
+        return record.pos - weight.position
+
+    def _check_biallelic(self, record):
+        if record is not None and len(record.alts) > 1:
+            raise RuntimeError('Only biallelic variants are supported: ' + str(record))
```
[the remark this answers is not in the document]
Comment (michaelgatzen) [reply ImputationPipeline/ScoreBGE/ScoreBGE.py:51]: Basically because of the `DS` dosage field. According to the header line
```
##FORMAT=<ID=DS,Number=1,Type=Float,Description="Genotype dosage">
```
it's a Number=1 field, which means that there is one entry per line. Some online research suggests that some tools may emit it as Number=A, which would mean one entry per allele, but I haven't found any clear standard of how that value would then be interpreted. Given that in this pipeline we only ever have biallelic imputed sites I figured it's not worth dealing with it. Let me know if you disagree @kachulis

File: ImputationPipeline/ScoreBGE/ScoreBGE.py [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+            self.gvcf_sites_scored[sample_name].append((weight.locus, weight.ref, weight.alt))
+            self.gvcf_sample_score[sample_name] += site_score
+    
+    def _vcf_score_dosage(self, record, weight):
+        for sample_name in self.sample_names:
+            if (weight.locus, weight.ref, weight.alt) in self.gvcf_sites_scored_set[sample_name]:
+                continue
+
+            if 'DS' not in record.samples[sample_name]:
+                raise RuntimeError('VCF file does not contain dosage information for sample ' + sample_name + ' at site ' + weight.locus + ':' + weight.ref + ':' + weight.alt)
```
[the remark this answers is not in the document]
Comment (michaelgatzen) [reply ImputationPipeline/ScoreBGE/ScoreBGE.py:76]: No good reason here, just assuming that we're using GLIMPSE2 imputed files, which always have DS. Can add that functionality

File: ImputationPipeline/ScoreBGE/ScoreBGE.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (kachulis) [reply ImputationPipeline/ScoreBGE/ScoreBGE.py:51]: got it, ok yeah that seems reasonable.

File: ImputationPipeline/ScoreBGE/tests/resources/wgs_vcf/test_wgs.vcf [hunk 15]
```diff
@@ -0,0 +1,18 @@
+##fileformat=VCFv4.2
+##FORMAT=<ID=DS,Number=1,Type=Float,Description="Genotype dosage">
+##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
+##FILTER=<ID=LowDepth,Description="Set if true:DP<=1">
+##contig=<ID=1,length=248956422>
+##contig=<ID=2,length=242193529>
+##contig=<ID=3,length=198295559>
+##reference=file:///broad/tag_working/edytam/dragen/reference_tables/hg38_hash_graph/reference.bin
```
Comment (kachulis) [inline ImputationPipeline/ScoreBGE/tests/resources/wgs_vcf/test_wgs.vcf:8]: remove

File: ImputationPipeline/ScoreBGE/tests/resources/combined/test_combined.wgs.vcf [hunk 16]
```diff
@@ -0,0 +1,12 @@
+##fileformat=VCFv4.2
+##FORMAT=<ID=DS,Number=1,Type=Float,Description="Genotype dosage">
+##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
+##contig=<ID=1,length=248956422>
+##contig=<ID=2,length=242193529>
+##contig=<ID=3,length=198295559>
+##reference=file:///broad/tag_working/edytam/dragen/reference_tables/hg38_hash_graph/reference.bin
```
Comment (kachulis) [inline ImputationPipeline/ScoreBGE/tests/resources/combined/test_combined.wgs.vcf:7]: remove

File: ImputationPipeline/ScoreBGE/tests/test_ScoreBGE.py [hunk 17]
```diff
@@ -0,0 +1,129 @@
...
+
+    def test_gvcf_score_variants(self):
+        s = ScoreBGE.BGEScorer('resources/ref.dict', 'resources/test_weights_gvcf_variants.txt')
+
+        s.score_wes_gvcf('resources/test_gvcf_variants.gvcf.gz', ['testsample'], 30)
+        expected_sites_scored = [
+            ('3:100', 'A', 'C'),  # score 2*1
+            ('3:100', 'A', 'C'),  # score 0
+            ('3:110', 'A', 'C'),  # score 4
+            ('3:110', 'A', 'C'),  # score 8
```
Comment (kachulis) [inline ImputationPipeline/ScoreBGE/tests/test_ScoreBGE.py:65]: oh nice, these are some tricky edge cases

File: ImputationPipeline/PCATasks.wdl [hunk 18]
```diff
@@ -114,21 +116,18 @@ task ArrayVcfToPlinkDataset {
...
-      --vcf ~{vcf} \
-      --extract-intersect ~{pruning_sites} ~{subset_to_sites} \
-      --allow-extra-chr \
-      --set-all-var-ids @:#:\$1:\$2 \
-      --new-id-max-allele-len 1000 missing \
-      --out ~{basename} \
-      --make-bed \
-      --rm-dup force-first
-  >>>
+  command {
```
Comment (kachulis) [inline ImputationPipeline/PCATasks.wdl:126]: can we leave these as `<<<` instead of `{`

File: ImputationPipeline/ScoreBGE/tests/resources/wes_gvcf_variants/test_gvcf_variants.gvcf [hunk 19]
```diff
@@ -0,0 +1,22 @@
+##fileformat=VCFv4.2
+##ALT=<ID=NON_REF,Description="Represents any possible alternative allele at this location">
+##FORMAT=<ID=GQ,Number=1,Type=Integer,Description="Genotype Quality">
+##FORMAT=<ID=GT,Number=1,Type=String,Description="Genotype">
+##INFO=<ID=END,Number=1,Type=Integer,Description="Stop position of the interval">
+##contig=<ID=1,length=248956422>
+##contig=<ID=2,length=242193529>
+##contig=<ID=3,length=198295559>
+##reference=file:///broad/tag_working/edytam/dragen/reference_tables/hg38_hash_graph/reference.bin
```
Comment (kachulis) [inline ImputationPipeline/ScoreBGE/tests/resources/wes_gvcf_variants/test_gvcf_variants.gvcf:9]: remove

**Review by kachulis — APPROVED** [review]
Thanks for all the tests!  Just a couple very small things and then good to go.
